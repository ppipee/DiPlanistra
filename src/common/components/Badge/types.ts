export enum BadgeVariant {
	Outlined = 'outlined',
	Filled = 'filled',
}

export type BadgeVariantTypes = BadgeVariant | 'filled' | 'outlined'

export interface BadgeProps {
	$variant?: BadgeVariantTypes
	$color: string
	$secondColor?: string
}
