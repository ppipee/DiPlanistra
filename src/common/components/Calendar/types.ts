export interface CalenderDateProps {
	startDate?: Date
	endDate?: Date
}

export interface CalendarDisplayStateProps {
	year: number
	month: number
}

export interface CalendarDisplayRange {
	startRange?: CalendarDisplayStateProps
	endRange?: CalendarDisplayStateProps
}

export interface CalendarCellProps {
	$inBetween?: boolean
	$selected?: boolean
	$variant?: CalendarCellVariant
	$isFirstCell?: boolean
	$isLastCell?: boolean
	$isDateSelectable?: boolean
}

export type CalendarCellVariant = 'currentMonth' | 'otherMonth' | 'currentDate'
