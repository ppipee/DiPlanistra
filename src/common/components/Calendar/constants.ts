import { main } from 'common/styles/colors'

export const DAYS = ['Su', 'Mo', 'Tu', 'We', 'Th', 'Fr', 'Sa']

export const CALENDAR_COLOR = {
	Active: main[500],
	Hover: '#7CC0C6',
	Between: '#BFE7DF',
}
