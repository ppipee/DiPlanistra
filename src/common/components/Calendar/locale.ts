export const CALENDAR_START_DATE_TITLE = ['เริ่มต้น', 'Start Date']
export const CALENDAR_END_DATE_TITLE = ['สิ้นสุด', 'End Date']
