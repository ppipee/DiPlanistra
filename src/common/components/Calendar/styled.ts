import styled from 'styled-components'

import spaces from 'common/styles/mixins/spaces'

export const CalendarContainer = styled.div`
	padding: ${spaces(4)};
	display: inline-block;
	width: 280px;
	user-select: none;
`
