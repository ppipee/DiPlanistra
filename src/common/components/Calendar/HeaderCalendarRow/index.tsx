import React, { memo } from 'react'

import { Cell, CellWrapper } from '../components/CalendarCell/styled'
import { RowContainer } from '../components/CalendarRow/styled'
import { DAYS } from '../constants'

const HeaderCalendarRow = () => {
	return (
		<RowContainer>
			{DAYS.map((day) => (
				<CellWrapper key={`header-${day}`}>
					<Cell className="heading4">{day}</Cell>
				</CellWrapper>
			))}
		</RowContainer>
	)
}

export default memo(HeaderCalendarRow)
