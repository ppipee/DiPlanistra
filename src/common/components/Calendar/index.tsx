import React, { ReactNode } from 'react'

import { CalendarDisplayRange, CalenderDateProps } from 'common/components/Calendar/types'

import { CalendarCellDate } from './components/CalendarCell'
import CalendarHeader from './components/CalendarHeader'
import CalendarProvider from './components/CalendarProvider'
import CalendarTable from './components/CalendarTable'
import { CalendarContainer } from './styled'

export type Props = {
	defaultDate?: CalenderDateProps
	setDate: (date: CalenderDateProps) => void
	isDateSelectable?: (date: CalendarCellDate) => boolean
	dateRangeShown?: CalendarDisplayRange
	children?: ReactNode
}

const Calendar = ({ defaultDate, setDate, isDateSelectable, dateRangeShown, children }: Props) => {
	return (
		<CalendarProvider
			isDateSelectable={isDateSelectable}
			defaultDate={defaultDate}
			setDate={setDate}
			dateRangeShown={dateRangeShown}
		>
			<CalendarContainer>
				<CalendarHeader />
				<CalendarTable />
				{children}
			</CalendarContainer>
		</CalendarProvider>
	)
}

export default Calendar
