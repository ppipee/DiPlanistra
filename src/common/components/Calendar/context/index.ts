import { createContext } from 'react'

import { CalendarCellDate } from '../components/CalendarCell'
import { CalenderDateProps, CalendarDisplayStateProps, CalendarDisplayRange } from '../types'

export interface CalendarContextType {
	date: CalenderDateProps
	setDate: (date: CalenderDateProps) => void
	displayState: CalendarDisplayStateProps
	setDisplayState: (state: CalendarDisplayStateProps) => void
	dateHover?: Date
	setDateHover: (date?: Date) => void
	isDateSelectable: (date: CalendarCellDate) => boolean
	dateRangeShown?: CalendarDisplayRange
}

const CalendarContext = createContext<CalendarContextType | null>(null)

export default CalendarContext
