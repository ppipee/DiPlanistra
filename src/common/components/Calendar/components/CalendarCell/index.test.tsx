import React from 'react'

import 'jest-styled-components'
import { render, fireEvent } from '@testing-library/react'

import { CalendarCellVariant } from 'common/components//Calendar/types'

import * as CalendarCellType from '.'

describe('Test CalendarCell', () => {
	let DEFAULT_DATE: Date
	const useCellStyleSpy = jest.fn()
	const useDateHoverSpy = jest.fn()
	const useSelectDateSpy = jest.fn()
	const useCalendarContextSpy = jest.fn()
	const selectDateSpy = jest.fn()
	const mouseEnterSpy = jest.fn()
	const mouseLeaveSpy = jest.fn()

	jest.doMock('common/components/Calendar/hooks/useCellStyle', () => useCellStyleSpy)
	jest.doMock('common/components/Calendar/hooks/useDateHover', () => useDateHoverSpy)
	jest.doMock('common/components/Calendar/hooks/useSelectDate', () => useSelectDateSpy)
	jest.doMock('common/components/Calendar/hooks/useCalendarContext', () => useCalendarContextSpy)

	const { default: CalendarCell } = require('.') as typeof CalendarCellType

	afterEach(() => {
		jest.resetAllMocks()
	})

	beforeEach(() => {
		DEFAULT_DATE = new Date(2020, 11, 1)
		useCellStyleSpy.mockReturnValue({
			$inBetween: false,
			$selected: false,
			$variant: 'currentMonth' as CalendarCellVariant,
			$isFirstCell: false,
			$isLastCell: false,
		})
		useSelectDateSpy.mockReturnValue(selectDateSpy)
		useDateHoverSpy.mockReturnValue({
			mouseEnter: mouseEnterSpy,
			mouseLeave: mouseLeaveSpy,
		})
		useCalendarContextSpy.mockReturnValue({
			isDateSelectable: () => true,
		})
	})

	it('should render CalendarCell with number value', () => {
		const { container } = render(<CalendarCell date={DEFAULT_DATE} index={1} />)

		expect(useCellStyleSpy).toBeCalledTimes(1)
		expect(useSelectDateSpy).toBeCalledTimes(1)
		expect(useDateHoverSpy).toBeCalledTimes(1)
		expect(useCalendarContextSpy).toBeCalledTimes(1)
		expect(container.firstChild).toMatchSnapshot()
	})

	it('should render CalendarCell with null value', () => {
		const { container } = render(<CalendarCell date={null} index={1} />)

		expect(useCellStyleSpy).toBeCalledTimes(1)
		expect(useSelectDateSpy).toBeCalledTimes(1)
		expect(useDateHoverSpy).toBeCalledTimes(1)
		expect(useCalendarContextSpy).toBeCalledTimes(1)
		expect(container.firstChild).toMatchSnapshot()
	})

	it('should render CalendarCell with hover and inBetween styled', () => {
		useCellStyleSpy.mockReturnValue({
			$inBetween: true,
			$selected: false,
			$variant: 'currentMonth' as CalendarCellVariant,
			$isFirstCell: false,
			$isLastCell: false,
		})

		const { container } = render(<CalendarCell date={DEFAULT_DATE} index={1} />)
		expect(container.firstChild).toMatchSnapshot()
	})

	it('should render CalendarCell with hover styled for first cell on row', () => {
		useCellStyleSpy.mockReturnValue({
			$inBetween: true,
			$selected: false,
			$variant: 'currentMonth' as CalendarCellVariant,
			$isFirstCell: true,
			$isLastCell: false,
		})

		const { container } = render(<CalendarCell date={DEFAULT_DATE} index={0} />)
		expect(container.firstChild).toMatchSnapshot()
	})

	it('should render CalendarCell with hover styled for last cell on row', () => {
		useCellStyleSpy.mockReturnValue({
			$inBetween: true,
			$selected: false,
			$variant: 'currentMonth' as CalendarCellVariant,
			$isFirstCell: false,
			$isLastCell: true,
		})

		const { container } = render(<CalendarCell date={DEFAULT_DATE} index={1} />)

		expect(container.firstChild).toMatchSnapshot()
	})

	it('should render CalendarCell with currentMonth variant', () => {
		useCellStyleSpy.mockReturnValue({
			$variant: 'currentMonth' as CalendarCellVariant,
		})

		const { container } = render(<CalendarCell date={DEFAULT_DATE} index={1} />)
		expect(container.firstChild).toMatchSnapshot()
	})

	it('should render CalendarCell with otherMonth variant', () => {
		useCellStyleSpy.mockReturnValue({
			$variant: 'otherMonth' as CalendarCellVariant,
		})

		const { container } = render(<CalendarCell date={DEFAULT_DATE} index={1} />)
		expect(container.firstChild).toMatchSnapshot()
	})

	it('should render CalendarCell with currentDate date variant', () => {
		useCellStyleSpy.mockReturnValue({
			$variant: 'currentDate' as CalendarCellVariant,
		})

		const { container } = render(<CalendarCell date={DEFAULT_DATE} index={1} />)
		expect(container.firstChild).toMatchSnapshot()
	})

	it('should perform onClick correctly', () => {
		const { container } = render(<CalendarCell date={DEFAULT_DATE} index={1} />)

		fireEvent.click(container.firstChild.firstChild)

		expect(selectDateSpy).toBeCalledTimes(1)
		expect(selectDateSpy).toBeCalledWith(DEFAULT_DATE)
	})

	it('should perform mouseEnter correctly', () => {
		const { container } = render(<CalendarCell date={DEFAULT_DATE} index={1} />)

		fireEvent.mouseEnter(container.firstChild.firstChild)

		expect(mouseEnterSpy).toBeCalledTimes(1)
		expect(mouseEnterSpy).toBeCalledWith(DEFAULT_DATE)
	})

	it('should perform mouseLeave correctly', () => {
		const { container } = render(<CalendarCell date={DEFAULT_DATE} index={1} />)

		fireEvent.mouseLeave(container.firstChild.firstChild)

		expect(mouseLeaveSpy).toBeCalledTimes(1)
	})
})
