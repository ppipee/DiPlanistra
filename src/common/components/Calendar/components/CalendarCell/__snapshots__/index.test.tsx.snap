// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test CalendarCell should render CalendarCell with currentDate date variant 1`] = `
.c1 {
  width: inherit;
  height: inherit;
  margin: auto;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-radius: 4px;
  position: absolute;
  -webkit-transition: all 0.1s ease-in;
  transition: all 0.1s ease-in;
  cursor: pointer;
  color: #00A600;
}

.c1:hover {
  background-color: #CEF4CE;
}

.c0 {
  width: 32px;
  height: 32px;
  position: relative;
  padding: 0 4px;
  box-sizing: content-box;
}

<div
  class="c0"
>
  <div
    class="c1 heading4"
  >
    1
  </div>
</div>
`;

exports[`Test CalendarCell should render CalendarCell with currentMonth variant 1`] = `
.c1 {
  width: inherit;
  height: inherit;
  margin: auto;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-radius: 4px;
  position: absolute;
  -webkit-transition: all 0.1s ease-in;
  transition: all 0.1s ease-in;
  cursor: pointer;
  color: #7D8391;
}

.c1:hover {
  background-color: #CEF4CE;
}

.c0 {
  width: 32px;
  height: 32px;
  position: relative;
  padding: 0 4px;
  box-sizing: content-box;
}

<div
  class="c0"
>
  <div
    class="c1 heading4"
  >
    1
  </div>
</div>
`;

exports[`Test CalendarCell should render CalendarCell with hover and inBetween styled 1`] = `
.c1 {
  width: inherit;
  height: inherit;
  margin: auto;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-radius: 4px;
  position: absolute;
  -webkit-transition: all 0.1s ease-in;
  transition: all 0.1s ease-in;
  cursor: pointer;
  color: #515A6C;
}

.c1:hover {
  background-color: #CEF4CE;
}

.c0 {
  width: 32px;
  height: 32px;
  position: relative;
  padding: 0 4px;
  box-sizing: content-box;
}

.c0::before {
  content: '';
  width: 100%;
  position: absolute;
  height: inherit;
  background-color: #EEFBEE;
}

<div
  class="c0"
>
  <div
    class="c1 heading4"
  >
    1
  </div>
</div>
`;

exports[`Test CalendarCell should render CalendarCell with hover styled for first cell on row 1`] = `
.c1 {
  width: inherit;
  height: inherit;
  margin: auto;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-radius: 4px;
  position: absolute;
  -webkit-transition: all 0.1s ease-in;
  transition: all 0.1s ease-in;
  cursor: pointer;
  color: #515A6C;
}

.c1:hover {
  background-color: #CEF4CE;
}

.c0 {
  width: 32px;
  height: 32px;
  position: relative;
  padding: 0 4px;
  box-sizing: content-box;
}

.c0::before {
  content: '';
  width: 100%;
  position: absolute;
  height: inherit;
  background-color: #EEFBEE;
  border-radius: 4px 0 0 4px;
}

<div
  class="c0"
>
  <div
    class="c1 heading4"
  >
    1
  </div>
</div>
`;

exports[`Test CalendarCell should render CalendarCell with hover styled for last cell on row 1`] = `
.c1 {
  width: inherit;
  height: inherit;
  margin: auto;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-radius: 4px;
  position: absolute;
  -webkit-transition: all 0.1s ease-in;
  transition: all 0.1s ease-in;
  cursor: pointer;
  color: #515A6C;
}

.c1:hover {
  background-color: #CEF4CE;
}

.c0 {
  width: 32px;
  height: 32px;
  position: relative;
  padding: 0 4px;
  box-sizing: content-box;
}

.c0::before {
  content: '';
  width: 100%;
  position: absolute;
  height: inherit;
  background-color: #EEFBEE;
  border-radius: 0 4px 4px 0;
  width: calc(100% - 8px);
}

<div
  class="c0"
>
  <div
    class="c1 heading4"
  >
    1
  </div>
</div>
`;

exports[`Test CalendarCell should render CalendarCell with null value 1`] = `
.c1 {
  width: inherit;
  height: inherit;
  margin: auto;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-radius: 4px;
  position: absolute;
  -webkit-transition: all 0.1s ease-in;
  transition: all 0.1s ease-in;
  cursor: pointer;
  color: #7D8391;
}

.c1:hover {
  background-color: #CEF4CE;
}

.c0 {
  width: 32px;
  height: 32px;
  position: relative;
  padding: 0 4px;
  box-sizing: content-box;
}

.c0 > div {
  pointer-events: none;
}

<div
  class="c0"
>
  <div
    class="c1 heading4"
  />
</div>
`;

exports[`Test CalendarCell should render CalendarCell with number value 1`] = `
.c1 {
  width: inherit;
  height: inherit;
  margin: auto;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-radius: 4px;
  position: absolute;
  -webkit-transition: all 0.1s ease-in;
  transition: all 0.1s ease-in;
  cursor: pointer;
  color: #7D8391;
}

.c1:hover {
  background-color: #CEF4CE;
}

.c0 {
  width: 32px;
  height: 32px;
  position: relative;
  padding: 0 4px;
  box-sizing: content-box;
}

<div
  class="c0"
>
  <div
    class="c1 heading4"
  >
    1
  </div>
</div>
`;

exports[`Test CalendarCell should render CalendarCell with otherMonth variant 1`] = `
.c1 {
  width: inherit;
  height: inherit;
  margin: auto;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-radius: 4px;
  position: absolute;
  -webkit-transition: all 0.1s ease-in;
  transition: all 0.1s ease-in;
  cursor: pointer;
  color: #9298A3;
}

.c1:hover {
  background-color: #CEF4CE;
}

.c0 {
  width: 32px;
  height: 32px;
  position: relative;
  padding: 0 4px;
  box-sizing: content-box;
}

<div
  class="c0"
>
  <div
    class="c1 heading4"
  >
    1
  </div>
</div>
`;
